### ISSUE TITLE
ObjectManager example

### ISSUE THREAD

**Issue by ddompke** [python-sdbus/python-sdbus#16]
Hi, are there any examples of correct usage of `DbusObjectManagerInterfaceAsync`? I tried inheriting it instead of `DbusInterfaceCommonAsync`, but I couldn't get it to work (there was no `org.freedesktop.Dbus.ObjectManager` interface while introspecting). Any help would be greatly appreciated.

**Comment by igo95862** [comment]
Do you want to access it as a client or export it to dbus as server?

**Comment by ddompke** [comment]
I want to export it to dbus as server.

**Comment by igo95862** [comment]
> I want to export it to dbus as server.

I did not finish this functionality at the moment. It needs extra code to automatically emit interfaces added and removed signals.

You can do a work around for now.

Copy this class: https://github.com/python-sdbus/python-sdbus/blob/1a8ab053c8f0044559be1ffa98e8a4c23b33cfec/src/sdbus/dbus_proxy_async.py#L906

and remove `serving_enabled=False,`

You will have to manually emit `interfaces_added` and `interfaces_removed` signals.

**Comment by igo95862** [comment]
> I am not sure if I understand correctly. Should I call the `export_to_dbus()` on the instance of modified `DbusObjectManagerInterfaceAsync`? Doing so causes the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/###/###/dbusDemo/service.py", line 40, in <module>
>     loop.run_until_complete(startup())
>   File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
>     return future.result()
>   File "/home/###/###/dbusDemo/service.py", line 32, in startup
>     obm.export_to_dbus('/')
>   File "/home/###/###/dbusDemo/venv/lib/python3.8/site-packages/sdbus/dbus_proxy_async.py", line 759, in export_to_dbus
>     bus.add_interface(new_interface, object_path,
> sd_bus_internals.SdBusLibraryError: Line: 365. sd_bus_add_object_vtable(self->sd_bus_ref, &interface_object->interface_slot->slot_ref, path_char_ptr, interface_name_char_ptr, interface_object->vtable, interface_object) in function SdBus_add_interface returned error: 22
> ```
> 
> When I change the `interface_name` of `DbusObjectManagerInterfaceAsync` to something different than `org.freedesktop.DBus.ObjectManager`, this error no longer occurs.

Looks like `sdbus` prevents exporting one of the reserved types.

**Comment by ddompke** [comment]
I see, thanks. I am afraid I can't offer much help with implementation as I don't have much experience with `sdbus`. But if you ever need somebody to test whether `DbusObjectManagerInterfaceAsync` works correctly, I am willing to help.

**Comment by ddompke** [comment]
Can you provide some rough estimate on how long it would take to finish? There is no hurry, but it would be nice to know in advance.

**Comment by igo95862** [comment]
@ddompke I start a branch [` object-manager-finish-wip `](https://github.com/python-sdbus/python-sdbus/tree/object-manager-finish-wip) which adds a new method `export_with_manager`.

See 72805801933ceb492b4f3cd3e0c177c7801c13da unit test on how to use. (will have documentation in final release)

(also no deletion so far)

**Comment by ddompke** [comment]
Thanks, tested it and it works great so far. I will give an update if I run into any issues.

**Comment by igo95862** [comment]
I merged the code and documentation for exporting objects with ObjectManager.

Have you ran in to any issues so far?

**Comment by ddompke** [comment]
After quick testing, I think that `remove_managed_object` doesn't quite work as intended. It does indeed remove object from tree, but `interfaces_removed` signal data seems incorrect. According to documentation, it should contain list of interfaces that were removed. Currently it lists all interfaces (plus `org.freedesktop.DBus.ObjectManager`, which was never part of the object) even if they were not removed. This happens when I export multiple objects under one path, and then delete one of them.

In my example `object` and `object2` were both imported under path `/objects/object1`, and `ObjectManager` was exported to `/`. I removed only `object`. I got the following signal:

```
Type=signal  Endian=l  Flags=1  Version=1  Priority=0 Cookie=8                                
  Sender=:1.181  Path=/  Interface=org.freedesktop.DBus.ObjectManager  Member=InterfacesRemove
d                                                                                             
  UniqueName=:1.181                                                                           
  MESSAGE "oas" {                                                                             
          OBJECT_PATH "/objects/object1";                                                     
          ARRAY "s" {                                                                         
                  STRING "org.freedesktop.DBus.Peer";                                         
                  STRING "org.freedesktop.DBus.Introspectable";                               
                  STRING "org.freedesktop.DBus.Properties";                                   
                  STRING "org.freedesktop.DBus.ObjectManager";                                
                  STRING "org.example.interface2";                                            
                  STRING "org.example.interface";                                             
          };                                                                                  
  };          
```

`org.example.interface2`, `org.freedesktop.DBus.Peer`, `org.freedesktop.DBus.Introspectable`, `org.freedesktop.DBus.Properties` were not actually removed, and `org.freedesktop.DBus.ObjectManager` was never in /objects/object1.

I expected a signal that was more along the lines of:

```
  Type=signal  Endian=l  Flags=1  Version=1  Priority=0 Cookie=8                                
  Sender=:1.181  Path=/  Interface=org.freedesktop.DBus.ObjectManager  Member=InterfacesRemoved                                                                                             
  UniqueName=:1.181                                                                           
  MESSAGE "oas" {                                                                             
          OBJECT_PATH "/objects/object1";                                                     
          ARRAY "s" {                                                                                                                                                
                  STRING "org.example.interface";                                             
          };                                                                                  
  };  
```

**Comment by ddompke** [comment]
Just to clarify, `object` implements interface `org.example.interface` and `object2` implements `org.example.interface2`.

**Comment by igo95862** [comment]
> This happens when I export multiple objects under one path

I don't think that supposed to be possible. I should probably make `object2` replace `object` when added to same path.

> plus `org.freedesktop.DBus.ObjectManager`

I also noticed that but it feels like an error in `sdbus`. I will try to investigate and report the bug to systemd.

**Comment by igo95862** [comment]
https://github.com/systemd/systemd/issues/8444

**Comment by igo95862** [comment]
@ddompke would you mind testing if `get_managed_objects` also adds `org.freedesktop.DBus.ObjectManager` to managed objects?

**Comment by ddompke** [comment]
Yes, `get_managed_objects` adds `org.freedesktop.DBus.ObjectManager` to managed objects? Now that I took a closer look, the signal `InterfacesAdded` also contains `org.freedesktop.DBus.ObjectManager`

**Comment by igo95862** [comment]
Ok. Thank you. I will see if that bug can be fixed upstream.

**Comment by igo95862** [comment]
I released version 0.10.0 which has the ObjectManager methods.

I also opened the draft of pull request to fix the object managers interface being returned for non ObjectManager objects.
